### PR TITLE
Adding keyvault service (only standard SKU) with key & secret (soft delete feature not supported)

### PR DESCRIPTION
**Pull request by redak-pragma**
This PR aims to provide support for KeyVault service only SKU Standard tier.

In some circumstances meaning custom ASH environment, soft deletion should be configurable.

Additionally, sometimes `object_id` is not a valid UUID.

### CODE REVIEW

File: internal/services/keyvault/key_vault_access_policy_resource.go [hunk 1]
```diff
@@ -0,0 +1,372 @@
+package keyvault
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2019-09-01/keyvault"
```
Comment (katbyte) [inline internal/services/keyvault/key_vault_access_policy_resource.go:10]: could we change this to use the profile version if it exists?

**Review by katbyte — CHANGES_REQUESTED** [review]
quick comment while passing through and checking tests:

image

File: internal/services/keyvault/key_vault_access_policy_resource.go (hunk 1 above)
Comment (redak-hash) [reply internal/services/keyvault/key_vault_access_policy_resource.go]: Indeed, profile `2020-09-01` exists and points to `2019-09-01` kv service. https://github.com/azure/azure-sdk-for-go/blob/v61.6.0/profiles/2020-09-01/keyvault/mgmt/keyvault/models.go#L15

done https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/fc43f67ad8dd2773422152654c58c75fc649b9e1

**Comment by redak-hash** [comment]
Regarding acceptance tests, I think that we have an unexpected behavior when soft delete feature is enabled.

Do we assume that, at this moment, soft deletion is not supported ?
I don't find any official documentation regarding this situation.

any opinion @katbyte @tombuildsstuff ?

Thank you.

**Comment by katbyte** [comment]
@redak-pragma - is this something that is not supported in azure stack at all? or a feature that needs to be enabled on a stack instance?

**Comment by redak-pragma** [comment]
> @redak-pragma - is this something that is not supported in azure stack at all? or a feature that needs to be enabled on a stack instance?

@katbyte , not supported at all

File: internal/provider/features.go [hunk 2]
```diff
@@ -10,6 +10,24 @@ func schemaFeatures() *pluginsdk.Schema {
...
+		// 				Type:     pluginsdk.TypeBool,
+		// 				Optional: true,
+		// 			},
+		// 			"purge_soft_delete_on_destroy": {
+		// 				Type:     pluginsdk.TypeBool,
+		// 				Optional: true,
+		// 			},
+		// 		},
+		// 	},
+		// },
```
Comment (katbyte) [inline internal/provider/features.go:30]: ditto for all unsupported stuff

File: website/docs/r/key_vault_secret.html.markdown [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+The following attributes are exported:
+
+* `id` - The Key Vault Secret ID.
+* `version` - The current version of the Key Vault Secret.
+
+## Timeouts
+
+
+
```
Comment (katbyte) [inline website/docs/r/key_vault_secret.html.markdown:91]: extra new lines here
```suggestion
```

File: website/docs/r/key_vault.html.markdown [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+* `enable_rbac_authorization` - (Optional) Boolean flag to specify whether Azure Key Vault uses Role Based Access Control (RBAC) for authorization of data actions. Defaults to `false`.
+
+* `network_acls` - (Optional) A `network_acls` block as defined below.
+
+~> **Note:** This field can only be configured one time and cannot be updated.
+
+<!-- NOTE: not supported -->
+<!-- * `contact` - (Optional) One or more `contact` block as defined below.
+
+~> **Note:** This field can only be set once user has `managecontacts` certificate permission. -->
```
Comment (katbyte) [inline website/docs/r/key_vault.html.markdown:100]: even docs nbot supported should be removed

File: internal/services/keyvault/key_vault_access_policy_data_source_test.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/tf/acceptance"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/tf/acceptance/check"
+)
+
+type KeyVaultAccessPolicyDataSource struct{}
+
+func TestAccDataSourceKeyVaultAccessPolicy_key(t *testing.T) {
```
Comment (katbyte) [inline internal/services/keyvault/key_vault_access_policy_data_source_test.go:14]: same for tests
```suggestion
func TestAccKeyVaultAccessPolicyDataSource_key(t *testing.T) {
```

File: internal/services/keyvault/key_vault_access_policy_resource.go [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/az/resourceid"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/locks"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/tf"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/tf/timeouts"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/utils"
+)
+
+func resourceKeyVaultAccessPolicy() *schema.Resource {
```
Comment (katbyte) [inline internal/services/keyvault/key_vault_access_policy_resource.go:24]: resource is default so
```suggestion
func keyVaultAccessPolicy() *schema.Resource {
```

File: internal/services/keyvault/key_vault_access_policy_resource.go [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+	return &schema.Resource{
+		Create: resourceKeyVaultAccessPolicyCreate,
+		Read:   resourceKeyVaultAccessPolicyRead,
+		Update: resourceKeyVaultAccessPolicyUpdate,
+		Delete: resourceKeyVaultAccessPolicyDelete,
```
Comment (katbyte) [inline internal/services/keyvault/key_vault_access_policy_resource.go:29]:
```suggestion
		Create: keyVaultAccessPolicyCreate,
		Read:   keyVaultAccessPolicyRead,
		Update: keyVaultAccessPolicyUpdate,
		Delete: keyVaultAccessPolicyDelete,
```

File: internal/services/keyvault/key_vault_data_source.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+	"github.com/hashicorp/terraform-provider-azurestack/internal/az/tags"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/services/keyvault/validate"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/tf/set"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/tf/timeouts"
+	"github.com/hashicorp/terraform-provider-azurestack/internal/utils"
+)
+
+func dataSourceKeyVault() *schema.Resource {
```
Comment (katbyte) [inline internal/services/keyvault/key_vault_data_source.go:20]:
```suggestion
func keyVaultDataSource() *schema.Resource {
```

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by katbyte** [comment]
Thanks @redak-pragma - looking better, still have some comments i left inline as well 4 failing tests to address:
image

File: internal/features/user_flags.go [hunk 9]
```diff
@@ -1,9 +1,16 @@
...
+	// KeyVault      KeyVaultFeatures
 	ResourceGroup ResourceGroupFeatures
 }
 
+// NOTE: soft delete feature not yet supported with ASH 2108
+// type KeyVaultFeatures struct {
+// 	PurgeSoftDeleteOnDestroy    bool
+// 	RecoverSoftDeletedKeyVaults bool
+// }
+
```
[the remark this answers is not in the document]
Comment (redak-pragma) [reply internal/features/user_flags.go]: https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/380ae1ae98407c69f3c3a0ad66a69464ba6b0b2b sure, removed

File: internal/provider/features.go (hunk 2 above)
Comment (redak-pragma) [reply internal/provider/features.go]: https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/380ae1ae98407c69f3c3a0ad66a69464ba6b0b2b, removed also

File: website/docs/r/key_vault_secret.html.markdown (hunk 3 above)
Comment (redak-pragma) [reply website/docs/r/key_vault_secret.html.markdown]: https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/380ae1ae98407c69f3c3a0ad66a69464ba6b0b2b cleaned

File: website/docs/r/key_vault.html.markdown (hunk 4 above)
Comment (redak-pragma) [reply website/docs/r/key_vault.html.markdown]: https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/380ae1ae98407c69f3c3a0ad66a69464ba6b0b2b cleaned

File: internal/services/keyvault/key_vault_access_policy_data_source.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+import (
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/2020-09-01/keyvault/mgmt/keyvault"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func dataSourceKeyVaultAccessPolicy() *schema.Resource {
```
[the remark this answers is not in the document]
Comment (redak-pragma) [reply internal/services/keyvault/key_vault_access_policy_data_source.go]: https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/380ae1ae98407c69f3c3a0ad66a69464ba6b0b2b sure, fixed

File: internal/services/keyvault/key_vault_access_policy_data_source_test.go (hunk 5 above)
Comment (redak-pragma) [reply internal/services/keyvault/key_vault_access_policy_data_source_test.go]: https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/380ae1ae98407c69f3c3a0ad66a69464ba6b0b2b fixed also

File: internal/services/keyvault/key_vault_access_policy_resource.go (hunk 7 above)
Comment (redak-pragma) [reply internal/services/keyvault/key_vault_access_policy_resource.go]: https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/380ae1ae98407c69f3c3a0ad66a69464ba6b0b2b fixed

File: internal/services/keyvault/key_vault_data_source.go (hunk 8 above)
Comment (redak-pragma) [reply internal/services/keyvault/key_vault_data_source.go]: https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/380ae1ae98407c69f3c3a0ad66a69464ba6b0b2b fixed

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @redak-pragma - tests pass now so overall this is good to go aside from needed to remove the commented out code. we don't just leave it in. once thats done i'll get this merged!

**Comment by redak-pragma** [comment]
> Thanks @redak-pragma - tests pass now so overall this is good to go aside from needed to remove the commented out code. we don't just leave it in. once thats done i'll get this merged!

@katbyte , thank you for your review,

Here my recent changes https://github.com/hashicorp/terraform-provider-azurestack/pull/151/commits/60f72bb9715b1ce956483db55f61c60cf2a71c3f (will be squashed as soon as you review again this PR, thank you in advance.)

**Review by katbyte — APPROVED** [review]
Thanks @redak-pragma - no need as i'll squash when i merge. LGTM now! 🦀

**Pull request merged**